backend/src/apiserver/resource/resource_manager.go

```diff
@@ -229,6 +233,56 @@ func (r *ResourceManager) GetPipelineTemplate(pipelineId string) ([]byte, error)
 	return template, nil
 }
 
+// user-gcp-sa secret was the default way to authenticate in GCP, but it is no longer needed when
+// we suggest using default authentication methods provided by GKE.
+// Alternatives include: project default service account, node pool default service account and
+// workload identity.
+// Dropping GOOGLE_APPLICATION_CREDENTIALS env when the secret is not configured ensures old
+// pipelines can still be run in new recommended clusters set up as above methods.
+//
+// TODO: when all pipeline samples no longer use user-gcp-sa secret, we can remove this usage.
+func (r *ResourceManager) dropUserGcpSaIfNotConfigured(workflow util.Workflow) {
+	someStepsHaveUserGcpSa := false
+	for _, template := range workflow.Workflow.Spec.Templates {
+		for _, volume := range template.Volumes {
+			if volume.Secret != nil && volume.Secret.SecretName == defaultGcpSecretName {
+				someStepsHaveUserGcpSa = true
+			}
+		}
+	}
+	if !someStepsHaveUserGcpSa {
+		// No steps used user-gcp-sa secret, no need to drop GOOGLE_APPLICATION_CREDENTIALS.
+		return
```

Comment:
would that possible: no steps/components mounted secret but set the ADC env?

Reply:
Let me explain. We only handle the default case of using use_gcp_secret with all default arguments.

If user doesn't have this secret, but have ADC env, they may have mounted secret as a different name. So we shouldn't break them